s3fs/core.py

```diff
@@ -142,7 +169,10 @@ def _ls(self, path, refresh=False):
                     f['Size'] = 0
                     del f['Name']
             else:
-                files = self.s3.list_objects(Bucket=bucket).get('Contents', [])
+                try:
+                    files = self.s3.list_objects(Bucket=bucket).get('Contents', [])
+                except ClientError:
+                    files = []
```

Comment:
- Is there a cheap way to determine the existence of a bucket?
A listing of buckets is not enough, as the path could be to a public bucket not associated with the user. Listing the objects ought to be slower than doing `head_bucket`, but probably not by much compared with the connection time; I can't tell the difference with `%time`.

Worth noting that if the following keys exist
bucket/file1
bucket/dir/dir/file2
then ls('bucket') will NOT give 'bucket/dir' or 'bucket/dir/dir', but perhaps it should. The trouble is that 'bucket/dir' can also be a file with data.